database/migrations/sqls/20240529120605-pgwar-entity-up.sql

```diff
@@ -0,0 +1,195 @@
+    time_span jsonb,
+    first_second bigint,
+    last_second bigint,
+    tmsp_last_modification timestamp with time zone,
+    CONSTRAINT entity_preview_pkey PRIMARY KEY (pk_entity, fk_project)
+)
+PARTITION BY LIST (fk_project);
+
+-- Create partition for community (fk_project = 0)
+--------------------------------------------------
+CREATE TABLE pgwar.entity_preview_0 PARTITION OF pgwar.entity_preview
+FOR VALUES IN (0);
+
+-- Function to create partition on pgwar.entity_preview
+-------------------------------------------------------
+CREATE FUNCTION pgwar.add_entity_preview_partition()
+    RETURNS TRIGGER
+    LANGUAGE plpgsql
+    AS $_$
+BEGIN
+    -- create the partition
+    EXECUTE format('CREATE TABLE pgwar.entity_preview_%1$s PARTITION OF pgwar.entity_preview FOR VALUES IN (%1$s);', NEW.pk_entity);
+    RETURN NEW;
+END;
+$_$;
+
+-- Trigger on projects.project to create new partition on pgwar.entity_preview
+------------------------------------------------------------------------------
+CREATE TRIGGER add_pgwar_entity_preview_partition
+    BEFORE INSERT ON projects.project
+    FOR EACH ROW
+    EXECUTE FUNCTION pgwar.add_entity_preview_partition();
+
+-- Function to upsert fk_project, pk_entity, fk_class on pgwar.entity_preview
+-----------------------------------------------------------------------------
+CREATE OR REPLACE FUNCTION pgwar.upsert_entity_preview_fk_class(entity_id int, project_id int, class_id int)
+    RETURNS VOID
+    AS $$
+BEGIN
+    INSERT INTO pgwar.entity_preview(pk_entity, fk_project, fk_class)
+        VALUES(entity_id, project_id, class_id)
+    ON CONFLICT(pk_entity, fk_project)
+        DO UPDATE SET
+            -- ... or update the fk_class
+            fk_class = EXCLUDED.fk_class
+        WHERE
+            -- ... where it is distinct from previous value
+            entity_preview.fk_class IS DISTINCT FROM EXCLUDED.fk_class;
+END;
+$$
+LANGUAGE plpgsql;
+
+-- Trigger function after_modify_info_proj_rel
+----------------------------------------------
+CREATE FUNCTION pgwar.after_modify_info_proj_rel()
+    RETURNS TRIGGER
+    LANGUAGE plpgsql
+    AS $$
+DECLARE
+    is_upsert boolean;
+    entity information.resource;
+BEGIN
+    -- get the referenced information.resource
+    SELECT
+        * INTO entity
+    FROM
+        information.resource
+    WHERE
+        pk_entity = COALESCE(NEW.fk_entity, OLD.fk_entity);
+    -- if the referenced item is an entity
+    IF entity.pk_entity IS NOT NULL THEN
+        -- determine if this is an upsert action
+        SELECT
+            (NEW.is_in_project = TRUE
+                AND TG_OP != 'DELETE') INTO is_upsert;
+        -- if upsert ...
+        IF is_upsert = TRUE THEN
+            -- ... upsert the project entity
+            PERFORM
+                pgwar.upsert_entity_preview_fk_class(NEW.fk_entity, NEW.fk_project, entity.fk_class);
+            -- if allowed ...
+            IF (entity.community_visibility ->> 'toolbox')::bool = TRUE THEN
```

Comment:
Could it be possible to have a single IF statement to check if it's in a project or in community, store the fk_project in a variable with a COALESCE(NEW.fk_project, OLD.fk_project, 0) (so it would take the project, or 0) and then only have one pgwar.upsert_entity_preview_fk_class() and one DELETE FROM pgwar.entity_preview ?
correct me if I'm wrong about the fk_project

Reply:
No, this would not implement the desired logic, because eventually want two entity_previews for one info_proj_rel-information.resource join:
One for the community version and one for the project version.

The desired logic in other words:

- We want one community entity_preview (with fk_project=0) for each information.resource that meets two requirements
  - the resource has community_visibility->'toolbox'=true
  - the resource is in at least one project (= it is referenced by one info_proj_rel having is_in_project=true)
- We want one project entity_preview (with fk_project=1..n) for each info_proj_rel having is_in_project=true and referencing an information.resource

Expressed in pseudo code (e=information.resource, ipr=projects.info_proj_rel, p=projects.project, ep=pgwar.entity_preview):

p1, p2, e1_true, ipr_p1_e1_true, ipr_p2_e1_true
--> ep_p0_e1, ep_p1_e1, ep_p2_e1

p1, p2, e1_true, ipr_p1_e1_true, ipr_p2_e1_false
--> ep_p0_e1, ep_p1_e1

p1, p2, e1_true, ipr_p1_e1_false, ipr_p2_e1_false
--> (no ep, because e1 in no project)

p1, p2, e1_false, ipr_p1_e1_true, ipr_p2_e1_true
--> ep_p1_e1, ep_p2_e1 (no ep_p0_e1 because community visibility is false, expressed as e1_false here)

Reply:
Ok, thanks for the clarification, now I understand!